### ISSUE TITLE
File carving of write open files

### ISSUE THREAD

**Issue by daniellu96** [osquery/osquery#6192]
# Bug report

### What operating system and version are you using?
OS - Windows 10.0.18363

### What version of osquery are you using?
osquery version - 4.1.2

### What steps did you take to reproduce the issue?
Tried using Zentral to carve files from the hosts

### What did you expect to see?
The carved file

### What did you see instead?
A file with size of 1K that contains only \x00

#### The Story
I tried to carve the windows security evtx but it failed so i tried to investigate the reason.
Eventually, it seems that osquery can't carve a file with open handle with write permission.
I tried doing the carving using Zentral.

Reproduce the error:
Carving a simple txt file worked
Trying to carve the same file while using the python open command on it with "w" or "a" flags (f = open(r"C:\path\to\file", flag) ) ended with a 1K file that contains only \x00.

Thanks

**Comment by mbhatt1** [comment]
You mean a locked file?

**Comment by alessandrogario** [comment]
Hello @daniellu96 

I believe this is expected behavior on Windows, and is caused by a combination of two things:
 * Mandatory file locking (and the defaults it imposes when creating file handles)
 * Sharing mode, as specified by a process when opening a file

In short: it is not normally possible to open a handle to a file that is already opened by another process that did not pass the [FILE_SHARE_READ](https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-createfilea) sharing mode flag to CreateFile.

I think that osquery should not return an empty file, and generate an error log instead

**Comment by daniellu96** [comment]
Hi, thanks for the replies :)

@mbhatt1
yes, exactly

@alessandrogario
Not sure about it, as I can open the same file for reading using 2 different process.
I love python, and used it to find this bug, and I can open the same text file with 2 python process, and notepad++ (in any order, so no one is blocking the other).

As i mentioned in the beginning, my purpose was to read windows logs, and i'm sure that's possible because i can use the powershell Get-EventLog cmdlet to get them

Maybe it should be a different functionality, because i just run procmon and it doesn't seem there that powershell opens the evtx file itself (?)

**Comment by alessandrogario** [comment]
> Hi, thanks for the replies :)
> 
> @mbhatt1
> yes, exactly
> 
> @alessandrogario
> Not sure about it, as I can open the same file for reading using 2 different process.
> I love python, and used it to find this bug, and I can open the same text file with 2 python process, and notepad++ (in any order, so no one is blocking the other).
> 
> As i mentioned in the beginning, my purpose was to read windows logs, and i'm sure that's possible because i can use the powershell Get-EventLog cmdlet to get them
> 
> Maybe it should be a different functionality, because i just run procmon and it doesn't seem there that powershell opens the evtx file itself (?)

I do not think that the Get-EventLog cmdlet opens the raw files directly; it is possible that it makes use of the Windows Event Log API to access the data (which means that it's not subject to the issues described above about mandatory file locking/sharing mode).

Testing with Notepad++ and Python is not enough, as it really depends on how the original process (a Windows component) has opened the file. The `FILE_SHARE_READ` flag description in the [CreateFile documentation](https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-createfilea) explains why kind of test is not reliable

**Comment by daniellu96** [comment]
Running api monitor shows that the Get-EventLog cmdlet runs different api - OpenEventLogW
Not sure that its completely different, as i suspect that OpenEventLogW will run NTCreatefile (Maybe i will check it later)

But about the locked file, i tested it also against a local text file i created, not only the windows event log which is managed by windows.

Checking procmon, python (the first process) opened the file with "Share mode : Read, Write", which by my understanding, should enable subsequent open operations on the file to request read and write access.

**Comment by daniellu96** [comment]
Why closed? The issue isn't solved yet and I'm sure that everyone would be happier with a solution as file carving is a major component in investigations.

**Comment by alessandrogario** [comment]
Hello @daniellu96!

The scope of the issue was not exactly clear, and in a standard situation it's normal that the same file can't be opened twice unless the first application opening the handle allows it.

Editors often open & close the file right away when reloading data, so the only way to reliably test this issue is to directly control the sharing flags of all the programs involved and also track how long the handles are kept alive on both processes.

The test should be performed with a simple file, without involving any other external resource (like the Windows Event Log). I have written the following test program in cpp that can be used to keep the file open, but Python can be used too if the exact call can be replicated:

```cpp
#ifdef UNICODE
#undef UNICODE
#endif

#include <Windows.h>
#include <iostream>

// (1) Make sure this file exists
const std::string kSampleFilePath{"C:\\test_file.txt"};

int main() {
  // (2) Change this value from `true` to `false` to test file locking
  bool share_file = true;

  auto file_handle = CreateFile(kSampleFilePath.c_str(),
                                GENERIC_READ,
                                share_file ? FILE_SHARE_READ : 0, // FILE_SHARE_WRITE can be added too
                                nullptr,
                                OPEN_EXISTING,
                                FILE_ATTRIBUTE_NORMAL,
                                nullptr);

  if (file_handle == INVALID_HANDLE_VALUE) {
    std::cerr << "Failed to open the handle\n";
    return 1;
  }

  std::cout << "Sleeping 5 seconds...\n";
  Sleep(5000);

  CloseHandle(file_handle);
  return 0;
}
```
osquery should be able to carve the file when `share_file` is set to true, and fail when it is set to `false`.

Producing an empty file when the handle can't be opened is a bug, as I think it should just log an error and skip it.

**Comment by theopolis** [comment]
> Producing an empty file when the handle can't be opened is a bug, as I think it should just log an error and skip it.

Agreed
